python/lsst/ts/nightreport/testutils.py

```diff
@@ -0,0 +1,468 @@
+        reports = await create_test_database(
+            postgres_url=postgresql_url,
+            num_reports=num_reports,
+            num_edited=num_edited,
+        )
+
+        dsn = dsn_from_connection_info(conn.info)
+        db_config = db_config_from_dsn(dsn)
+        with modify_environ(
+            SITE_ID=TEST_SITE_ID,
+            **db_config,
+        ):
+            # Note: httpx.AsyncClient does not trigger startup and shutdown
+            # events. We could use asgi-lifespan's LifespanManager,
+            # but it does not trigger the shutdown event if there is
+            # an exception, so it does not seem worth the bother.
+            assert not shared_state.has_shared_state()
+            await main.startup_event()
+            try:
+                async with httpx.AsyncClient(
+                    app=main.app, base_url="http://test"
+                ) as client:
+                    assert shared_state.has_shared_state()
+                    yield client, reports
+            finally:
+                await main.shutdown_event()
+
+
+@contextlib.contextmanager
+def modify_environ(**kwargs: typing.Any) -> collections.abc.Iterator:
+    """Context manager to temporarily patch os.environ.
+
+    This calls `unittest.mock.patch` and is only intended for unit tests.
+
+    Parameters
+    ----------
+    kwargs : `dict` [`str`, `str` or `None`]
+        Environment variables to set or clear.
+        Each key is the name of an environment variable (with correct case);
+        it need not already exist. Each value must be one of:
+
+        * A string value to set the env variable.
+        * None to delete the env variable, if present.
+
+    Raises
+    ------
+    RuntimeError
+        If any value in kwargs is not of type `str` or `None`.
+
+    Notes
+    -----
+    Example of use::
+
+        ...
+        def test_foo(self):
+            set_value = "Value for $ENV_TO_SET"
+            with modify_environ(
+                HOME=None,  # Delete this env var
+                ENV_TO_SET=set_value,  # Set this env var
+            ):
+                self.assertNotIn("HOME", os.environ)
+                self.assert(os.environ["ENV_TO_SET"], set_value)
+    """
+    bad_value_strs = [
+        f"{name}: {value!r}"
+        for name, value in kwargs.items()
+        if not isinstance(value, str) and value is not None
+    ]
+    if bad_value_strs:
+        raise RuntimeError(
+            "The following arguments are not of type str or None: "
+            + ", ".join(bad_value_strs)
+        )
+
+    new_environ = os.environ.copy()
+    for name, value in kwargs.items():
+        if value is None:
+            new_environ.pop(name, None)
+        else:
+            new_environ[name] = value
+    with unittest.mock.patch("os.environ", new_environ):
+        yield
+
+
+def assert_good_response(response: httpx.Response) -> typing.Any:
+    """Assert that a response is good and return the data.
```

Comment:
Please add an explanation for the return type to the doc string here and everywhere else.